### PR TITLE
feature(commitlog): Checking commitlog in sct

### PR DESCRIPTION
**Pull request by temichus**
Checking "free segments" and "commitlog directory" metrics in SCT task: https://github.com/scylladb/qa-tasks/issues/1523

introdus new SCT background thread CommitLogCheckThread

if commitlog-use-hard-size-limit is enabled,
this thread will check the following metrics during the test:

1) commit log directory does not exceed the limit at any stage of the test. if it exceeded - send an error event that fails the test.

2) free segments don't drop to zero at any stage of the test, otherwise, send an error event that fails the test.

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: sdcm/commit_log_check_thread.py
Comment (fruch) [inline sdcm/commit_log_check_thread.py:178]: please remove the warnings if not needed

File: defaults/test_default.yaml
Comment (fruch) [inline defaults/test_default.yaml:237]: by default we enable the thread, but the feature is enabled only on one test, wouldn't it gonna fail on any other test ?

File: test-cases/longevity/longevity-10gb-3h.yaml
Comment (fruch) [inline test-cases/longevity/longevity-10gb-3h.yaml:25]: is this feature gonna be default in scylla ? do we want it default in SCT ?

Comment (temichus) [reply test-cases/longevity/longevity-10gb-3h.yaml:25]: yes, it will be default in the future https://github.com/scylladb/scylladb/pull/10974.

But I changed here just for testing. will remove when PR will be ready

File: defaults/test_default.yaml
Comment (temichus) [reply defaults/test_default.yaml:237]: feature will be enabled by default Scylla https://github.com/scylladb/scylladb/pull/10974

iIdealy it should work in all tests. but this line is made to remove it by default, instead of reverting all code. or disable in some tests if needed 

scilla req: The commitlog didn't go over the size limit. It is a little more involved, we should make sure that the size of the commitlog directory never exceeds the limit, we can't rely on metrics since if there is a bug in the metrics we will never catch an actual bug where the directory is larger than the stats and the size limit was actually breached.

File: sdcm/commit_log_check_thread.py
Comment (temichus) [reply sdcm/commit_log_check_thread.py:57]: because returning  any value means an error - data will be returned per shard:

```
{'metric': {'__name__': 'scylla_commitlog_disk_total_bytes', 'instance': '10.4.3.68', 'job': 'manager_agent', 'shard': '0'}, 'values': [[1698765005.833, '8757706752'], [1698765006.833, '8757706752']..............
```

**Comment by temichus** [comment]
testing runs : 
1) special cluster where the thread shows errors   https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/artsiom_mishuta/job/longevity-10gb-3h-test-pr-5797/109/

2) longevity-10gb-3h with commitlog-use-hard-size-limit=true (failed but thread works correctly) https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/artsiom_mishuta/job/longevity-10gb-3h-test-pr-5797/110/

**Comment by temichus** [comment]
@scylladb/qa-maintainers,  PR is tested and ready for review. Can be merged now  because the thread started only if  commitlog-use-hard-size-limit is enabled (now  there are 0 such tests ) or postponed and merged after https://github.com/scylladb/scylladb/pull/10974 

But for sure we should present this functionality to all QAs before merge because it will affect all tests in SCT

**Review by soyacz — COMMENTED** [review]
quite late comment, but I wonder if we couldn't achieve this with `PrometheusAlertManagerListener` and setup proper alerts in prometheus.
Maybe worth to consider it for future (unless we can't use it).

Besides, small comments to fix and overall LGTM

File: sdcm/commit_log_check_thread.py
Comment (temichus) [reply sdcm/commit_log_check_thread.py]: done

File: sdcm/commit_log_check_thread.py
Comment (temichus) [reply sdcm/commit_log_check_thread.py]: done

File: sdcm/commit_log_check_thread.py
Comment (wmitros) [inline sdcm/commit_log_check_thread.py:71]: I'm not that familiar with prometheus queries format but for the `scylla_commitlog_disk_total_bytes>{commitlog_total_space_in_mb}` part we should have a `>=` (it doesn't matter now because in most default configurations `commitlog_total_space_in_mb` is not actually a multiple of segment size)

File: sdcm/commit_log_check_thread.py
Comment (wmitros) [inline sdcm/commit_log_check_thread.py:87]:
```suggestion
        this condition will ignore result when Scylla does not exceed  segments limit
```

**Review by wmitros — COMMENTED** [review]
I checked the invariants of the commitlog that we're checking and only found minor issues. However, recently I found that the method for getting the commitlog limit (based on whether the config option is -1) is not actually precise, so instead of hard-coding all the adjustments here I just posted https://github.com/scylladb/scylladb/pull/16020 which should allow getting the exact limit directly from the metrics (new metric will be named `scylla_commitlog_max_disk_size`). We may have to postpone merging this until that one's merged as well, but maybe it's not a problem as we're also waiting for https://github.com/scylladb/scylladb/pull/10974

File: sdcm/commit_log_check_thread.py
Comment (temichus) [reply sdcm/commit_log_check_thread.py]: fixed

File: sdcm/commit_log_check_thread.py
Comment (temichus) [reply sdcm/commit_log_check_thread.py]: fixed

**Comment by temichus** [comment]
Moved to draft 

waiting for https://github.com/scylladb/scylladb/pull/16020
blocked by: https://github.com/scylladb/scylladb/pull/10974

**Comment by temichus** [comment]
@wmitros I've updated PR to use /commitlog/metrics/max_disk_size API
Could you please check that the calculation is right?

test run with new code version https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/artsiom_mishuta/job/longevity-10gb-3h-test-pr-5797/117/

File: sdcm/commit_log_check_thread.py
Comment (wmitros) [inline sdcm/commit_log_check_thread.py:60]: If we want to be absolutely precise, we should have actually have a `>=` here (I might have missed it in the review of previous version)

File: sdcm/commit_log_check_thread.py
Comment (wmitros) [inline sdcm/commit_log_check_thread.py:20]: If we continue with the approach with the API, this is probably not necessary anymore

**Review by wmitros — APPROVED** [review]
LGTM, I found only 1 place where we weren't exactly precise. As we discussed, it seems like the approach without the API was good too, so if you prefer we can revert to that one. Otherwise we can just go ahead with this one

File: sdcm/commit_log_check_thread.py
Comment (temichus) [reply sdcm/commit_log_check_thread.py]: fixed

File: sdcm/commit_log_check_thread.py
Comment (temichus) [reply sdcm/commit_log_check_thread.py]: fixed

**Comment by temichus** [comment]
@soyacz, @fruch - could you please re review?  this PR ready to be merged.

File: sdcm/commit_log_check_thread.py
Comment (fruch) [inline sdcm/commit_log_check_thread.py:163]: can you explain this, it's not clear what this loop is doing...

Comment (fruch) [reply sdcm/commit_log_check_thread.py:163]: if I read it correctly it works every minute ? isn't that a bit too excessive ?

Comment (fruch) [reply sdcm/commit_log_check_thread.py:163]: the whole play with saving clocktime, and playing with it's clear

why not just wait the interval ?

Comment (temichus) [reply sdcm/commit_log_check_thread.py:163]: this was put to be able exit from thread every 60s(when test finished and termination_event is set) not every 600s 

but according to [doc](https://docs.python.org/3/library/threading.html#event-objects) it should work also as 
 self.termination_event.wait(self.check_interval) will test it

Comment (fruch) [reply sdcm/commit_log_check_thread.py]: you can see all the other places we have a temnication_event and a loop, like in Nemesis thread, that we do exactly that

Comment (temichus) [reply sdcm/commit_log_check_thread.py]: updated

**Review by fruch — APPROVED** [review]
LGTM

**Review by soyacz — APPROVED** [review]
LGTM

**Comment by soyacz** [comment]
@temichus please verify for next few weeks if works as expected and ping me for backport

**Comment by temichus** [comment]
@soyacz @fruch FYI
For backport it requires 2 Scylla PRs 
1)https://github.com/scylladb/scylladb/pull/10974
2) https://github.com/scylladb/scylladb/pull/16020

**Comment by yaronkaikov** [comment]
@temichus it seems that this commit is breaking the artifact tests. see here for example https://jenkins.scylladb.com/job/scylla-master/job/artifacts/job/artifacts-ubuntu2204-test/374/consoleFull

**Comment by fruch** [comment]
I'm reverting this until fixed
```
02:02:11  CommitLogCheckThread_:
02:02:11  Traceback (most recent call last):
02:02:11    File "/usr/local/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
02:02:11      self.run()
02:02:11    File "/usr/local/lib/python3.10/threading.py", line 953, in run
02:02:11      self._target(*self._args, **self._kwargs)
02:02:11    File "/tmp/workspace/scylla-master/artifacts/artifacts-ubuntu2204-test/scylla-cluster-tests/sdcm/commit_log_check_thread.py", line 162, in run
02:02:11      self.overflow_commit_log_directory_checker(self.start_time, interval_end_time)
02:02:11    File "/tmp/workspace/scylla-master/artifacts/artifacts-ubuntu2204-test/scylla-cluster-tests/sdcm/commit_log_check_thread.py", line 174, in overflow_commit_log_directory_checker
02:02:11      response = self.prometheus.query(
```

**Pull request merged**
